### ISSUE TITLE
Find / FindRef: wrapper function to find element based on ref

### ISSUE THREAD

**Issue by matt-oconnell** [vuejs/vue-test-utils#67]
**Suggestion for additional functionality**

Currently, there is no functionality built into `vue-test-utils` to select elements within a component based on `$refs` and return them within a Wrapper. Our test suite is built around the idea that we can easily tag elements we need to hook onto within tests using `$ref` rather than some arbitrary "marker class."

I think it might be helpful to include a specific method on the wrapper like `findRefs`

I'm not sure if this is a widely used strategy but we've found it helpful. Template structures change often and class selectors can easily break or target the wrong element when these changes happen.

**Comment by eddyerburgh** [comment]
I think this is a good idea, but I don't think we should add an extra method that only finds refs.

We could extend find, so that it takes an object. 

```js
find({
  selector: 'refs',
  value: 'value' 
})
```

Then we could extend it to other selectors, like `props`.

```js
find({
  selector: 'props',
  value: 'value' 
})
```

**Comment by matt-oconnell** [comment]
I wonder how many alternate use cases we would have for a selector other than CSS, component name, or $ref. I'm not sure if component with prop: `x`would ever be used as a selector, but I could be wrong.

**Comment by matt-oconnell** [comment]
But, if we do want to go with the selector/value find/findAll functions, I could start a PR. Initially it could look for $props / $refs. I guess the tough part would be matching `$props` because the values could be `any`

**Comment by eddyerburgh** [comment]
I think we should just add support for refs at the moment ☺

**Comment by matt-oconnell** [comment]
@eddyerburgh To make it a little more concise, how do you feel about:

```js
find({ ref: 'myRefName' })
```

**Comment by eddyerburgh** [comment]
Yep that looks good 👍

**Comment by matt-oconnell** [comment]
Great, updating the PR to reflect this. Currently, the validation error messages for find/findAll display: `wrapper.find/findAll() must be passed a valid CSS selector or a Vue constructor`. As we update the API to accept this new options object, how should we refer to it? 

If we're going to start with refs for now, we could use:

`wrapper.find() must be passed a valid CSS selector, Vue constructor, or ref object`

Or `wrapper.find() must be passed a valid CSS selector, Vue constructor, or options object`

My only concern with the latter is that it may be too vague.

**Comment by eddyerburgh** [comment]
Ah that's a good point.

If we add this then we need to:
- Update all methods that take a selector: `is`, `contains`, `find`, `findAll`
- Update the selectors page to include the refs object
- Link to the selectors page in the validation error

That's a fairly big change. Before we go further, I would like some input from others. 

**Do we want to extend the selector to include { refs: 'refName' }**

**Comment by matt-oconnell** [comment]
So, one thing other thing to note: When selecting $refs, we rely on the $ref property to be available on the wrapper.vm, because we want to limit the results to this scope. This is one of the benefits of $refs. If we are using a wrapper on a non-Vue HTML Element, we should not allow selecting by $ref.

**Comment by eddyerburgh** [comment]
Ok, we can add a validation error for that

**Comment by skray** [comment]
The PR for this change mentioned wanting some feedback from the community, so I thought I'd chime in.

This would be great to have, I liked the initial suggestion of having `wrapper.ref`, but I understand from a consistency standpoint why you'd want to extend the current finders. The current recommendation of using `wrapper.find({refs: 'refName'})` feels comfortable to me.

**Comment by eddyerburgh** [comment]
OK, I'm happy for this work to continue. @matt-oconnell do you have any blockers?

**Comment by matt-oconnell** [comment]
This is pretty much fully implemented in the corresponding [PR](https://github.com/vuejs/vue-test-utils/pull/68). The blocker was that the Flow types were invalid (specifically `Component`,) so this PR fails the type check unless `any` is used

**Comment by eddyerburgh** [comment]
I think we'll have to use any until proper flowtypes are exported from Vue — https://github.com/vuejs/vue/pull/5027

**Comment by eddyerburgh** [comment]
This can be closed, released in `1.0.0-beta.5`
